### PR TITLE
fix: Fix pact-matchers for nested objects

### LINKED ISSUE #58: Nested objects in interfaces/types are overwritten by last @pact-matcher used
When api function return type/interface contains nested object which itself contains one of the @pact-matcher - whole nested object in body in generated pact is replaced with the last pact-matcher defined. Example definitions and output:
```
export interface PactMatchersNestedType {
    nestedObject: {
        /** @pact-matcher hex */
        matcher?: string;
    }
}

/**
 * @pact
 * @pact-method GET
 * @pact-path /api
 */
export function apiNestedFunction(): PactMatchersNestedType {
    return {
        nestedObject: {}
    };
}
```

generated interaction looks like this:
```
      "description": "apiNestedFunction",
      "request": Object {
        "headers": Object {},
        "method": "GET",
        "path": "/api",
      },
      "response": Object {
        "body": Object {
          "nestedObject": "3F",
        },
        "headers": Object {},
        "matchingRules": Object {
          "$.body": Object {
            "match": "type",
          },
          "$.body.nestedObject": Object {
            "match": "regex",
            "regex": "^[0-9a-fA-F]+$",
          },
        },
        "status": 200,
      },
```

where as you can see **nestedObject** property inside body is replaced with hex from defined pact-matcher.

I've tested a few pact-matchers and this issue happens every time. If more than one pact-matcher is defined then the last one is used. 
Marked property can be optional or required, nested object can contain additional optional or required properties not marked with any matcher - with all this cases same issue occurs. 

Same issue happens for interface and TS type definition.

### PR DESCRIPTION
**Pull request by pawfa**
Fixes #58

### CODE REVIEW

**Review by crpc — APPROVED** [review]

File: tests/int/__snapshots__/pactGenerating.test.ts.snap [hunk 1]
```diff
@@ -662,6 +662,437 @@ Object {
...
+            "match": "regex",
+            "regex": "^[0-9a-f]{8}(-[0-9a-f]{4}){3}-[0-9a-f]{12}$",
+          },
+          "$.body.objectInline.matcher": Object {
+            "match": "regex",
+            "regex": "^[0-9a-fA-F]+$",
+          },
+          "$.body.objectInline.nestedArray": Object {
+            "match": "regex",
+            "regex": "^[0-9a-fA-F]+$",
```
Comment (Fiszcz) [inline tests/int/__snapshots__/pactGenerating.test.ts.snap:990]: This is incorrect matching rule.

`objectInline.nesgtedArray` is not string field, but array, so regex is incorrect

File: src/core/type-representation.ts [hunk 2]
```diff
@@ -99,7 +99,11 @@ const getObjectTypeRepresentation = (entryType: tsMorph.Type, source: tsMorph.So
 
         const propertyType = property.getTypeAtLocation(source);
         const propertyTypeRepresentation = getTypeRepresentation(propertyType, source);
-        if (typeMatcher && exampleRepresentation) {
+
+        // recursively set type representation for nested object properties
+        if (propertyType.isObject() && !propertyType.isArray()) {
```
Comment (Fiszcz) [inline src/core/type-representation.ts:104]: It is not solution to this problem.

The problem is in line #90 https://github.com/HLTech/pact-gen-ts/pull/60/files#diff-c6a2f1cd0cfa639da00a7d6c3dfb23df5d133f76e94823ec2619cd05535aa917R90 of this file. We get all descendants of object property (so, if AST contains some JSDocTag node for nested field of some parent field, this is persisted only for parent object field). So, the solution is getting only children nodes of JSDocComments kind of current field.

Working solution:
```ts
        const jsDocTagsOfProperty = property.getValueDeclaration()?.getChildrenOfKind(ts.SyntaxKind.JSDocComment).reduce((jsDocTags, jsDocCommentElement) => {
            return [...jsDocTags, ...(jsDocCommentElement.getChildrenOfKind(ts.SyntaxKind.JSDocTag))];
        }, []);
```
use it, and remove your code :)

File: tests/int/__snapshots__/pactGenerating.test.ts.snap (hunk 1 above)
Comment (Fiszcz) [reply tests/int/__snapshots__/pactGenerating.test.ts.snap:990]: I think you have created a too complex test case for the test, which is too large and finally, there are difficulties to analyze a lot of snapshot lines.

**Review by Fiszcz — CHANGES_REQUESTED** [review]
Fix things from my comment.

File: tests/int/pactGenerating.test.ts [hunk 3]
```diff
@@ -25,6 +24,16 @@ describe('createPacts', () => {
         });
     });
 
+    test('pact-matcher', () => {
```
Comment (Fiszcz) [inline tests/int/pactGenerating.test.ts:27]: Is it purposeful or some test residues?

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "pact-gen-ts",
-    "version": "0.9.0",
+    "version": "0.9.1",
```
Comment (Fiszcz) [inline package.json:3]: You have to update version to 0.9.2.

There has been already 0.9.1 version

**Comment by Fiszcz** [comment]
Looks nice

**Review by Fiszcz — APPROVED** [review]

File: package.json (hunk 4 above)
Comment (pawfa) [reply package.json:3]: Semantic-versioning library is super convenient :) - I guess we could think about implementing it.

File: tests/int/pactGenerating.test.ts (hunk 3 above)
Comment (pawfa) [reply tests/int/pactGenerating.test.ts:27]: Actually after consideration - I reverted this change in this PR - lets do it all at once in other refactor task - but the idea stays the same.

**Pull request merged**